### PR TITLE
Compound datatypes may not have members of size 0

### PR DESCRIPTION
**Pull request by e4t**
A member size of 0 may lead to an FPE later on as reported in CVE-2021-46244. To avoid this, check for this as soon as the member is decoded.
This should probably be done in H5O_dtype_decode_helper() already, however it is not clear whether all sizes are expected to be != 0.

This fixes CVE-2021-46244 / Bug #2242.

Signed-off-by: Egbert Eich <eich@suse.com>

### CODE REVIEW

**Comment by byrnHDF** [comment]
I think there should be release notes for these PRs.
It should go in the bug fixes/library section, with the newest dates at the top of the section.
The pattern is one line description, detailed summary and sig line of from (initials - date - issue #)

See other entries for examples.

**Comment by byrnHDF** [comment]
I think your first comment would be good.

- This fixes https://github.com/advisories/GHSA-vrxh-5gxg-rmhm
- 
  A member size of 0 may lead to an FPE later on as reported in https://github.com/advisories/GHSA-vrxh-5gxg-rmhm. To avoid 
  this, check for this as soon as the member is decoded.
  This should probably be done in H5O_dtype_decode_helper() already, however it is not clear whether all sizes are expected 
  to be != 0.

 (XXX - yyyy/mm/dd GH-2242)

File: src/H5Odtype.c [hunk 1]
```diff
@@ -332,6 +332,12 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                     H5MM_xfree(dt->shared->u.compnd.memb);
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "unable to decode member type")
                 } /* end if */
+                if (temp_type->shared->size == 0) {
+                    for (j = 0; j <= i; j++)
+                        H5MM_xfree(dt->shared->u.compnd.memb[j].name);
+                  H5MM_xfree(dt->shared->u.compnd.memb);
+                  HGOTO_ERROR(H5E_OHDR, H5E_BADVALUE, FAIL, "invalid field size in member type")
+                 }
```
Comment (lrknox) [inline src/H5Odtype.c]:
```suggestion
                    H5MM_xfree(dt->shared->u.compnd.memb);
                    HGOTO_ERROR(H5E_OHDR, H5E_BADVALUE, FAIL, "invalid field size in member type")
                }
```

Comment (lrknox) [reply src/H5Odtype.c]: These indentation changes should make clang-format checks pass when committed.

Comment (e4t) [reply src/H5Odtype.c]: Yeah, I've just noticed the formatting issue. I just fixed this and am about to resubmit.

@byrnHDF Regarding release notes - I haven't included  anything, yet, as I've submitted all fixes independently. Merging these would create conflicts in the release notes.
Would you perfer to have me submit release notes with each submission or should I submit them once the patch set has been merged?
I can go back into every fix, add release notes and resubmit.

**Comment by bmribler** [comment]
I think the CVE issue# should also be mentioned in the RELEASE.txt entry.

**Comment by lrknox** [comment]
> I think the CVE issue# should also be mentioned in the RELEASE.txt entry.

Yes, please include the CVE issue#

File: src/H5Odtype.c (hunk 1 above)
Comment (lrknox) [reply src/H5Odtype.c]: We would prefer that each PR has it's release note.  We can sort out the conflicts as we merge them.  Thanks!

**Comment by e4t** [comment]
> > I think the CVE issue# should also be mentioned in the RELEASE.txt entry.
> 
> Yes, please include the CVE issue#

Would you like me to add a RELEASE.txt entry to each patch I've pull-requested so far or should I do a joint patch when the patches have been merged?
The former would be cleaner and help associate the entry with the patch, however it would cause conflicts during merging which have to be resolved manually.

**Comment by lrknox** [comment]
Yes, please add a RELEASE.txt entry to each patch.  We can take care of resolving conflicts in RELEASE.txt when we merge the PRs.

**Comment by e4t** [comment]
> Yes, please add a RELEASE.txt entry to each patch. We can take care of resolving conflicts in RELEASE.txt when we merge the PRs.
Will do. Thanks!

I would also suggest to use the CVE IDs exclusively - instead of Github's IDs. The CVE IDs are resolved nicely on Github so there is not advantage using the Github IDs. Also CVE IDs are well established, thus people will understand them even when seeing them outside of a Github context.

**Comment by lrknox** [comment]
We need the Github Ids for internal purposes, but do want the CVE IDs for external users.

**Comment by e4t** [comment]
@ lrknox - I've created release note entries for 3 of the submissions and pushed them. Please have a look if they are Ok.
It's been a long day, I will do the remaining ones tomorrow. 
Thanks!

File: release_docs/RELEASE.txt [hunk 2]
```diff
@@ -142,14 +142,23 @@ New Features
 
 Support for new platforms, languages and compilers
 ==================================================
-    -
+    -  CVE-2021-46244 / GHSA-vrxh-5gxg-rmhm / Bug #2242
```
Comment (lrknox) [inline release_docs/RELEASE.txt]:
```suggestion
    -  CVE-2021-46244 / GHSA-vrxh-5gxg-rmhm
```

File: release_docs/RELEASE.txt [hunk 3]
```diff
@@ -142,14 +142,23 @@ New Features
...
+
+       Compound datatypes may not have members of size 0
+ 
+       A member size of 0 may lead to an FPE later on as reported in
+       CVE-2021-46244. To avoid this, check for this as soon as the
+       member is decoded.
+       This should probably be done in H5O_dtype_decode_helper() already,
+       however it is not clear whether all sizes are expected to be != 0.
 
+       (EFE - 2022/10/05 GEH-2243)
```
Comment (lrknox) [inline release_docs/RELEASE.txt]:
```suggestion
       (EFE - 2022/10/05 GH-2242)
```

**Review by lrknox — APPROVED** [review]

**Comment by lrknox** [comment]
> @ lrknox - I've created release note entries for 3 of the submissions and pushed them. Please have a look if they are Ok. It's been a long day, I will do the remaining ones tomorrow. Thanks!

@e4t  - Reviewed several of them and left suggestions per our release note conventions.  Thank you for all your work on these!

**Comment by e4t** [comment]
> > @ lrknox - I've created release note entries for 3 of the submissions and pushed them. Please have a look if they are Ok. It's been a long day, I will do the remaining ones tomorrow. Thanks!
> 
> @e4t - Reviewed several of them and left suggestions per our release note conventions. Thank you for all your work on these!

Thanks for reviewing! I've incorporated all the changes and pushed again. Hope I got everything ok, now!
Thank you for your patience!

**Comment by lrknox** [comment]
Our testing with h5format_convert and h5dump shows the FPE is gone, but we now have an "infinite loop closing library message.  we've seen this occasionally, but need to fix it before merging.

File: src/H5Odtype.c [hunk 4]
```diff
@@ -332,6 +332,12 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                     H5MM_xfree(dt->shared->u.compnd.memb);
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "unable to decode member type")
                 } /* end if */
+                if (temp_type->shared->size == 0) {
+                    for (j = 0; j <= i; j++)
+                        H5MM_xfree(dt->shared->u.compnd.memb[j].name);
```
Comment (fortnern) [inline src/H5Odtype.c]: Maybe make a macro or function for freeing incomplete compounds, these loops are repeated in multiple places (and should probably be added to the other errors in the loop as well)

**Comment by e4t** [comment]
> Our testing with h5format_convert and h5dump shows the FPE is gone, but we now have an "infinite loop closing library message. we've seen this occasionally, but need to fix it before merging.

Yeah, I recall seeing this - when I created the bug report and pull request I didn't remember to mention this. It is probably another problem triggered by the reproducer as the program no longer crashes immediately.
I may be able to take a look on Friday.

File: src/H5Odtype.c (hunk 4 above)
Comment (fortnern) [reply src/H5Odtype.c]: In general, especially given the reported failures, it is worth going over this section of code to make sure everything is cleaned up properly on error. I don't think temp_type is freed, or that any of the member types are freed. There may be other missing items too.

**Comment by fortnern** [comment]
I've pushed some changes that should clean up error recovery in this function. It probably needs to run through the formatter (not sure if that happens automatically in this fork)

**Review by fortnern — APPROVED** [review]

**Comment by lrknox** [comment]
h5dump and h5format_convert now fail appropriately with the test file POC-CVE-2021-46244.

**Pull request merged**